### PR TITLE
[PPC] Add custom lowering for uaddo

### PR DESCRIPTION
**Pull request by syzaara**
Improve the codegen for uaddo node for i64 in 64-bit mode and i32 in 32-bit mode by custom lowering.

### CODE REVIEW

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 1]
```diff
@@ -11967,11 +11970,51 @@ SDValue PPCTargetLowering::LowerFP_EXTEND(SDValue Op, SelectionDAG &DAG) const {
...
+    return SDValue();
+
+  EVT VT = Op.getNode()->getValueType(0);
+  bool is64Bit = Subtarget.isPPC64();
+
+  SDValue ADDC;
+  SDValue Overflow;
+  SDVTList VTs = Op.getNode()->getVTList();
+
+  ADDC = SDValue(DAG.getMachineNode(is64Bit ? PPC::ADDC8 : PPC::ADDC, dl, VT,
```
Comment (efriedma-quic) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:11995]: Please avoid getMachineNode() during legalization; add appropriate PPCISD nodes if necessary.  (Or maybe you can just use ISD::ADDC/ISD::ADDE.)

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 2]
```diff
@@ -198,6 +198,9 @@ PPCTargetLowering::PPCTargetLowering(const PPCTargetMachine &TM,
     }
   }
 
+  if (!Subtarget.hasP10Vector())
```
Comment (efriedma-quic) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:201]: How is hasP10Vector() related to the instructions in question?  Do we have testcases for what's generated when hasP10Vector() is enabled?

**Review by efriedma-quic — COMMENTED** [review]
Looks fine from my side, but I don't really have much context for PPC instruction sets, so a PPC expert should also look at it.

File: llvm/test/CodeGen/PowerPC/uaddo-32.ll [hunk 3]
```diff
@@ -0,0 +1,37 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=powerpc-unknown-linux-gnu | FileCheck %s
```
Comment (maryammo) [inline llvm/test/CodeGen/PowerPC/uaddo-32.ll:2]: Is the AIX assembly exactly the same? Can we add a RUN line for AIX?

File: llvm/test/CodeGen/PowerPC/uaddo-64.ll [hunk 4]
```diff
@@ -0,0 +1,37 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=powerpc64-unknown-linux-gnu | FileCheck %s
```
Comment (maryammo) [inline llvm/test/CodeGen/PowerPC/uaddo-64.ll:2]: Same here.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 5]
```diff
@@ -11967,11 +11969,45 @@ SDValue PPCTargetLowering::LowerFP_EXTEND(SDValue Op, SelectionDAG &DAG) const {
...
+  return Res;
+}
+
 /// LowerOperation - Provide custom lowering hooks for some operations.
 ///
 SDValue PPCTargetLowering::LowerOperation(SDValue Op, SelectionDAG &DAG) const {
   switch (Op.getOpcode()) {
   default: llvm_unreachable("Wasn't expecting to be able to lower this!");
+  case ISD::UADDO:
+    return LowerUaddo(Op, DAG);
```
Comment (lei137) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:12010]: Maybe we should keep the same formating as the rest of the case statements?

File: llvm/test/CodeGen/PowerPC/uaddo-32.ll [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+; RUN: llc < %s -mtriple=powerpc-ibm-aix-xcoff | FileCheck %s
+
```
Comment (lei137) [inline llvm/test/CodeGen/PowerPC/uaddo-32.ll:4]: Are these IR not valid for 64bit?  I would think we can just have 1 tc and make sure we test for both 32 and 64bit on Linux and AIX.

**Review by lei137 — COMMENTED** [review]
Is it possible to do an NFC patch to add in the new testcase so we can see the code gen improvements in this patch?

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 7]
```diff
@@ -11967,11 +11969,45 @@ SDValue PPCTargetLowering::LowerFP_EXTEND(SDValue Op, SelectionDAG &DAG) const {
   llvm_unreachable("ERROR:Should return for all cases within swtich.");
 }
 
+SDValue PPCTargetLowering::LowerUaddo(SDValue Op, SelectionDAG &DAG) const {
+  // Default to target independent lowering if there is a logical user of the
+  // carry-bit.
+  for (SDNode *U : Op->uses()) {
+    if (U->getOpcode() == ISD::SELECT || ISD::isBitwiseLogicOp(U->getOpcode()))
+      return SDValue();
```
Comment (maryammo) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:11977]: This prevents the following case to get custom lowered, however with custom lowering the final assembly looks better. There are 2 uaddo's which are used by logical or in the dag. 

```
unsigned long long add(unsigned long long a, unsigned long long b, unsigned long long c, unsigned long long *ovf)
{
     return __builtin_addcll(a,b,c,ovf);
}
```

$ clang -O3 -mcpu=power8 -m64 -S 
Assembly with default target independent lowering:
```
.add:
# %bb.0:                                # %entry
        add 4, 3, 4
        subc    3, 4, 3
        subfe 3, 4, 4
        neg 7, 3
        add 3, 4, 5
        subc    4, 3, 4
        subfe 4, 3, 3
        neg 4, 4
        or 4, 7, 4
        std 4, 0(6)
        blr 
```
Assembly with custom lowering:
```
.add:
# %bb.0:                                # %entry
        li 7, 0
        addc 3, 3, 4
        addze 4, 7
        addc 3, 3, 5
        addze 5, 7
        or 4, 4, 5
        std 4, 0(6)
        blr 
```

Comment (RolandF77) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:11977]: IIUC this patch doesn't make the long long case worse, it just doesn't make it better. Maybe long long can be addressed in a separate patch? Seems hard to tell that case apart from the condition use case - may want to handle the full 64 bit operation in 32 bit mode for long long.

Comment (maryammo) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:11977]: Yes, it does not make the long long case worse, it custom lowers long long in 64 bit mode except for some cases that returns to the default lowering.

**Review by maryammo — APPROVED** [review]
LGTM.

**Pull request merged**